listenbrainz_spark/recommendations/train_models.py

```diff
@@ -84,41 +175,95 @@ def train(training_data, validation_data, num_validation, ranks, lambdas, iterat
     for rank, lmbda, iteration in itertools.product(ranks, lambdas, iterations):
+        model_id = generate_model_id()
+
         t0 = time()
-        model_id = 'listenbrainz-recommendation-model-{}'.format(uuid.uuid4())
-        try:
-            model = ALS.trainImplicit(training_data, rank, iterations=iteration, lambda_=lmbda, alpha=alpha)
-        except Py4JJavaError as err:
-            current_app.logger.error('Unable to train model "{}"\n{}'.format(model_id, str(err.java_exception)), exc_info=True)
-            sys.exit(-1)
+        model = train(training_data, rank, iteration, lmbda, alpha, model_id)
         mt = '{:.2f}'.format((time() - t0) / 60)
+
         t0 = time()
-        try:
-            validation_rmse = compute_rmse(model, validation_data, num_validation)
-        except Py4JJavaError as err:
-            current_app.logger.error('Root Mean Squared Error for model "{}" for validation data not computed\n{}'.format(
-                model_id, str(err.java_exception)), exc_info=True)
-            sys.exit(-1)
+        validation_rmse = compute_rmse(model, validation_data, num_validation, model_id)
         vt = '{:.2f}'.format((time() - t0) / 60)
-        model_metadata.append((model_id, mt, rank, lmbda, iteration, "%.2f" % (validation_rmse), vt))
-        if best_model is None or validation_rmse < best_model.error:
-            best_model = Model(model=model, error=validation_rmse, rank=rank, lmbda=lmbda, iteration=iteration,
-                model_id=model_id, training_time=mt, rmse_time=vt)
 
-    best_model_metadata = {'error': '{:.2f}'.format(best_model.error), 'rank': best_model.rank, 'lmbda':
-            best_model.lmbda, 'iteration': best_model.iteration, 'model_id': best_model.model_id, 'training_time':
-                best_model.training_time, 'rmse_time': best_model.rmse_time}
-    return best_model, model_metadata, best_model_metadata
+        model_metadata.append((model_id, mt, rank, '{:.1f}'.format(lmbda), iteration, round(validation_rmse, 2), vt))
+
+        if best_model is None or validation_rmse < best_model.validation_rmse:
+            best_model = Model(
+                model=model,
+                validation_rmse=round(validation_rmse, 2),
+                rank=rank,
+                lmbda=lmbda,
+                iteration=iteration,
+                model_id=model_id,
+                training_time=mt,
+                rmse_time=vt,
+                alpha=alpha,
+            )
+
+    return best_model, model_metadata
+
+
+def delete_best_model():
+    """ Delete best model.
+        Note: At any point in time, only one model is in HDFS
+    """
+    dir_exists = utils.path_exists(path.DATA_DIR)
+    if dir_exists:
+        utils.delete_dir(path.DATA_DIR, recursive=True)
+
+
+def save_best_model(model_id, model):
+    """ Save best model to HDFS.
+
+        Args:
+            model_id (str): Best model identification string.
+            model: Trained best model
+    """
+    delete_best_model()
```

Comment:
Why is this here? This seems out of place for this function...

Reply:
I forgot add a comment. Previous model should be deleted before saving the new one.